### PR TITLE
make funcs argument type from torch.cuda.stream as torch.Stream

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
1. we want to support fsdp for custom device, so we make funcs argument type from torch.cuda.stream as torch.Stream

### CODE REVIEW

File: torch/distributed/fsdp/_utils.py [hunk 1]
```diff
@@ -54,9 +54,9 @@ def _same_storage_as_data_ptr(x: torch.Tensor, data_ptr: int) -> bool:
...
 
 
-def _no_dispatch_record_stream(tensor: torch.Tensor, stream: torch.cuda.Stream) -> None:
+def _no_dispatch_record_stream(tensor: torch.Tensor, stream: torch.Stream) -> None:
     # FIXME record_stream doesn't work with non-cuda tensors
     if not tensor.is_cuda:
         return
     with no_dispatch():
-        tensor.record_stream(cast(torch._C.Stream, stream))
+        tensor.record_stream(cast(torch._C.Stream, stream))  # type: ignore[redundant-cast]
```
Comment (awgu) [inline torch/distributed/fsdp/_utils.py:62]: curious: Can this `# type: ignore[redundant-cast]` be addressed somehow? For example, do we still need the `cast`?

Comment (heidongxianhua) [reply torch/distributed/fsdp/_utils.py]: > curious: Can this `# type: ignore[redundant-cast]` be addressed somehow? For example, do we still need the `cast`?

yes, I make some test, it works well when remove the `cast`

File: torch/distributed/fsdp/_utils.py [hunk 2]
```diff
@@ -54,9 +54,9 @@ def _same_storage_as_data_ptr(x: torch.Tensor, data_ptr: int) -> bool:
     return x._typed_storage()._data_ptr() == data_ptr
 
 
-def _no_dispatch_record_stream(tensor: torch.Tensor, stream: torch.cuda.Stream) -> None:
+def _no_dispatch_record_stream(tensor: torch.Tensor, stream: torch.Stream) -> None:
     # FIXME record_stream doesn't work with non-cuda tensors
-    if not tensor.is_cuda:
+    if tensor.device.type not in ["cuda", torch._C._get_privateuse1_backend_name()]:
```
[the remark this answers is not in the document]
Comment (heidongxianhua) [reply torch/distributed/fsdp/_utils.py:59]: > `torch._C._get_privateuse1_backend_name()` should have minimal overhead?

yeah, this func is just to get a  static string from c++, so the overhead is very small.

**Review by awgu — APPROVED** [review]
LGTM!

**Pull request closed without merging**